### PR TITLE
SPI touchscreen driver for MKS robin & Longer3D TFT

### PR DESCRIPTION
**Pull request by tpruvot**
follow up of PR #14578 deleted branch by mistake

### CODE REVIEW

**Comment by tpruvot** [comment]
remains one file change to change for the optional DMA

**Comment by tpruvot** [comment]
dogm buttons give such result : https://drive.google.com/file/d/1mFSWBuxq2iQ7tvqgWMIbzw2ETUNDmwz4/view?usp=sharing

i think i'm done

**Comment by thinkyhead** [comment]
Great! I have some screens here that I need to sort through. One of them might be similar to this.

File: Marlin/src/pins/stm32/pins_MKS_ROBIN_MINI.h [hunk 1]
```diff
@@ -112,7 +112,7 @@
 #define LCD_BACKLIGHT_PIN  PD13
 #define FSMC_CS_PIN        PD7    // NE4
 #define FSMC_RS_PIN        PD11   // A0
-#define TOUCH_CS           PC2
+#define TOUCH_CS_PIN       PC2
```
Comment (thinkyhead) [inline Marlin/src/pins/stm32/pins_MKS_ROBIN_MINI.h]: We could add `#if ENABLED(TOUCH_BUTTONS)` to the robin pins, if we think other LCDs might be used with these boards in the future.

**Comment by thisiskeithb** [comment]
If this could be enabled for the Robin Nano & it’s TFT, Tevo Nereus owners would be thrilled.

Those boards come with a locked down/closed source version of Marlin and the TFT becomes useless input-wise once you update to bugfix-2.0.x.

**Comment by thinkyhead** [comment]
This won't work yet. `HAS_GRAPHICAL_LCD` is **false** for `TOUCH_BUTTONS`. Therefore the U8G upscale code won't compile.

Pins files should wrap their definitions of `FSMC_CS` in `#if ENABLED(TOUCH_BUTTONS)` to ensure it doesn't get defined when some other graphical controller is enabled.

**Comment by tpruvot** [comment]
ok, so now i can do a PR for the longer3D STM32F103VE board.. Was the last really missing feature

**Comment by zillarob** [comment]
If you guys need something tested on a Robin nano (vet6), just let me know.

**Comment by tpruvot** [comment]
calibration values for the mks touchscreen could be different.. see https://github.com/makerbase-mks/MKS-Robin/blob/master/MKS%20Robin%20Nano/firmware/Marlin2.0-MKS-Robin_nano/Marlin/src/HAL/HAL_STM32F1/xpt2046.h#L40

**Comment by zillarob** [comment]
Thanks for the heads up! Ill poke at it a lil and see how see goes =D

**Comment by zillarob** [comment]
Im getting closer, but the buttons do not respond to touch.
I did enter the calibration values in the link above for the Robin tft.
Have Robin_tft and Touch_Buttons enabled



mar

**Comment by thinkyhead** [comment]
So for now I guess this can be a sub-option of `MKS_ROBIN_TFT` until more display models start to use touch.

Right now the DOGM code is all built around a 128x64 LCD, but new implementation files can certainly be added for larger displays, showing more information with the additional space.

**Comment by tpruvot** [comment]
thanks @zillarob nice to see the dogm part is fine on another board..

else for calibration, in fact current default should match MKS v2.0 one too, we are not at +/-10 on these values... its more a full reversed Y or so. Try clicking on the top part... only the red button will do something (enter menu) on status screen.

You could try to reverse mosi & miso pins numbers also... happen to be reversed when you play with DI/DO references in schematics

**Comment by zillarob** [comment]
So far I have tried all the calibrations. I do touch all over the screen to see if I get lucky =D

I have added the miso/mosi/sck/int to the pins file. Did notice that miso/mosi looked swapped so tried both ways. (I just noticed last night that MKS added the schematics, good on them)
Something that caught my eye is I had to add these to compile on your touchscreen version, but on the touchscreen_dogm version I think it compiled with or without them. 

I think these are more a Nano pins file thing, but Ive been having to do my best hatchet job on 2 things here:
If I have LCD_RESET_PIN defined I get: Marlin\src\lcd\dogm\../../inc/../pins/stm32/pins_MKS_ROBIN_NANO.h:111:28: error: 'PF6' was not declared in this scope. 
Not sure the f103vet6 has a PF6 pin? My best guess is PC6 would line up, but I have just been commenting it out and not using it.

If E0_AUTOFAN is defined, I get: 
Marlin\src\HAL\HAL_STM32F1\../../inc/SanityCheck.h:1385:8: error: #error "You cannot set E0_AUTO_FAN_PIN equal to FAN_PIN."
These pins are definitely set differently so not sure what is up here, other than it is above my paygrade. Ive been doing my best rusty hatchet job here also and commenting out the error in sanity check. Its dirty, but fans seem to work properly =D

Ill keep poking at it and see if I get lucky, prob start over fresh again today.
New to this github stuff so prob going to be a while before I figure out how to help you guys more, assuming my use of a rusty hatchet can be of use lol

**Comment by zillarob** [comment]
Started fresh and ran into a not in scope error.
Guessed that the " ; " needed to be a " , " and seemed to clear it up.

semi

**Comment by zillarob** [comment]
Ok, I think my eyes are going to start bleeding and need to take a break lol
Still not having any luck getting the buttons to respond to touch on the Nano.
With a little creative hatcheting, the Nano works with this one below.
Is there any way to see what pins they are are using in here for the spi stuff? Not sure if they are doing it differently or if it will have some clues in there, but I cant find them.
https://github.com/berghauz/Marlin-bugfix-2.0.x-FLSUN-QQ-S

**Comment by tpruvot** [comment]
rebased the branch without squashing... and fixed the error reported by @zillarob

**Comment by zillarob** [comment]
@tpruvot 

Looking at the Robin Nano, it seems int pin from the tft is nc. 
Im still struggling to get the touch to work and wondering if this might be the prob?

ETA  Metered from pin 11 on the hr2046 (penirq) to pin 26 on the ribbon cable connector on the Nano and get a circuit, but that seems to be where it stops. I cant find it from pin 26 there to the stm32.

tft

**Comment by tpruvot** [comment]
hmm, that may explain the "Z" coordinate probe i seen in mks code...

**Comment by tpruvot** [comment]
@zillarob could you try PR #14640

**Comment by zillarob** [comment]
@tpruvot  You are a rockstar!

I could give it a few swings before going to work and seems to be working.
I think its going to be a slow day so might be able to pop in and out later.
Here is the pins file that looked good in my quick testing this morning.

[pins_MKS_ROBIN_NANO.zip](https://github.com/MarlinFirmware/Marlin/files/3397150/pins_MKS_ROBIN_NANO.zip)

**Comment by thisiskeithb** [comment]
I have a Tevo Nereus config ready for a PR, so hopefully this works well. Great work @tpruvot & @zillarob!

**Comment by zillarob** [comment]
@thisiskeithb 

Careful with that pins file. I just pounded some defines out of the longer pins file in there with the hatchet. Im sure it could use some cleaning up =D

**Comment by zillarob** [comment]
@tpruvot 

Been playing around more this eve and is working well. I think you nailed it! Great job!!

**Pull request merged**